ENH: Implement overlap-add convolution

toddrjen opened the pull request:
This adds a new function `scipy.signal.oaconvolve`.  It implements the overlap-add method of convolution.  This approach is faster than `scipy.signal.convolve` when the arrays are large and faster than `scipy.signal.fftconvolve` when their sizes are significantly different.  Like `fftconvolve`, this version supports any subset of any number of dimensions.

Most overlap-add implementations I have seen use a loop-based approach, where they loop over each overlapping block.  The approach here uses a vectorized approach, where the input array is reshaped to fit the blocks and the FFT is applied to the entire pair of arrays at once.  This can have a big performance win for FFT implementations where plan creation is expensive.  Although this is less of an issue for the current default pocketfft implementation, last I heard support for multiple FFT backends is planned and many other FFT implementations have much more expensive plan creation. 

This approach will automatically fall back on using `fftconvolve` when the overlap-add method doesn't work, but it doesn't generally fall back when `fftconvolve` is faster.

Closes #1364

This also fixes a bug in `scipy.signal.fftconvolve` where `"valid"` mode convolution along a subset of dimensions can be incorrectly flagged as invalid.

Closes #10873

rgommers commented:
Nice! Obvious first question: can't this simply be added as `method='overlap-add'` to `signal.convolve`?

toddrjen commented:
Below is an initial comparison of performance between the new implementation and `fftconvolve`.  Note that these are log-log plots.  The red line represents the point where the two inputs are the same size.  The overlap-add approach is generally faster when the two inputs are different in size and neither array is very small.

![conv_comparison](https://user-images.githubusercontent.com/2272878/65567693-2344b480-df25-11e9-903b-dbbb5b13730a.png)

The next figure is similar, except the x axis is the ratio of sizes between the first array and second array.  

![conv_comparison_ratio](https://user-images.githubusercontent.com/2272878/65567772-5f781500-df25-11e9-9ddb-bda8d18c7925.png)

toddrjen commented:
> Nice! Obvious first question: can't this simply be added as `method='overlap-add'` to `signal.convolve`?

This was [discussed on the mailing list](https://mail.python.org/pipermail/scipy-dev/2019-June/023555.html) and the consensus seemed to be that a new function was preferable.  It has significantly different performance characteristics.  Ideally it would be folded into `convolve`, but that requires having a reliable way of determining the performance in a general way.  This isn't as easy as with `fftconvolve` due to how the implementation works.

mreineck commented:
> The approach here uses a vectorized approach, where the input array is reshaped to fit the blocks and the FFT is applied to the entire pair of arrays at once.

Great! This is also a big performance win with `pocketfft`, since it vectorizes multi-D transforms.

larsoner commented on scipy/signal/signaltools.py:
```diff
@@ -431,6 +442,329 @@ def fftconvolve(in1, in2, mode="full", axes=None):
+    .. [1] Wikipedia, "Overlap-add_method".
+           https://en.wikipedia.org/wiki/Overlap-add_method
+
+    .. versionadded:: 1.4.0
```
Should go in `Notes` section

larsoner commented on scipy/signal/signaltools.py:
```diff
@@ -33,7 +33,8 @@
 
-__all__ = ['correlate', 'fftconvolve', 'convolve', 'convolve2d', 'correlate2d',
+__all__ = ['correlate', 'correlate2d',
+           'convolve', 'convolve2d', 'fftconvolve', 'oaconvolve',
```
CircleCI complains because of ref problems, probably because you need to add this to `scipy/signal/__init__.py` itself as an import, and in the top docstring in the `Convolution` section

toddrjen replied:
I think this is fixed

larsoner replied:
I see no changes to `scipy/signal/__init__.py` in the diff, did you forget to push?

toddrjen replied:
I just pushed it

toddrjen replied:
Travis seems okay now.

ewmoore commented on scipy/signal/signaltools.py:
```diff
@@ -395,36 +469,334 @@ def fftconvolve(in1, in2, mode="full", axes=None):
+    # in an exception if the function falls back on `fftconvolve`.  This is
+    # to guarantee that oaconvolve function itself, rather than the
+    # fftconvolve fallback mode, is being used in tests and benchmarks
+    oa_only_error = ValueError('Using fftconvolve with_oa_only==True')
```
Do we really want to do this?  I'm not a big fan of a supposedly hidden argument that is actually exposed in the function definition and then therefore in the documentation.  Is there any precedence in Scipy for this kind of thing?

toddrjen replied:
I've replaced this with a normal argument.

toddrjen replied:
I have eliminated this entirely by using pytest's monkeypatch as discussed below.

toddrjen commented:
All checks pass and I think I have implemented all the suggestions so far.

larsoner commented:
Based on mailing list feedback, it's probably worth getting rid of the `method` / `oa_only_error` and using something else for testing purposes. I would personally do it with something like:
```
def fftconvolve_err(*args, **kwargs):
    raise RuntimeError('Fell back to fftconvolve')

...

def test_whatever(monkeypatch):
    monkeypatch.setattr(scipy.signal.signaltools, 'fftconvolve', fftconvolve_err)
    ...
```
It has local scope and it seems pretty readable.

toddrjen commented:
@larsoner 

> Based on mailing list feedback, it's probably worth getting rid of the `method` / `oa_only_error` and using something else for testing purposes. I would personally do it with something like:
> 
> ```
> def fftconvolve_err(*args, **kwargs):
>     raise RuntimeError('Fell back to fftconvolve')
> 
> ...
> 
> def test_whatever(monkeypatch):
>     monkeypatch.setattr(scipy.signal.signaltools, 'fftconvolve', fftconvolve_err)
>     ...
> ```
> 
> It has local scope and it seems pretty readable.

I agree. I have removed the argument and implemented the monkeypatching as suggested.  All the tests still pass.

toddrjen commented on scipy/signal/signaltools.py:
```diff
@@ -431,6 +489,311 @@ def fftconvolve(in1, in2, mode="full", axes=None):
+           where the overlap-add method isn't appropriate.
+        ``overlap-add``
+           Only use the overlap-add method.  Raise an exception in situations
+           where the overlap-add method isn't appropriate.
```
Sorry, I forgot to push the latest version

toddrjen commented:
I have pushed the version with all of the suggestions so far.  The travis failure appears to be an issue installing ubuntu rather than with my code.

toddrjen commented:
I have pushed the version with all of the suggestions so far. 

The travis failure appears to be an issue installing ubuntu rather than with my code.

peterbell10 commented:
@toddrjen could you try remaking your benchmark plots to make the comparison a bit clearer? As it is currently, both lines look almost on top of each other and the scales aren't very obvious. Maybe try plotting the ratio of `oaconvolve` time to `fftconvolve` time, or the difference? Sharing the benchmarks would be good as well.

Also, I don't know how others feel but squashing all of your changes into a single commit is making it harder for me to follow along with this PR. I've been meaning to leave a review, but when I come back the code seems to have changed significantly without any history.
If you're worried about the git history, my suggestion would be to submit 17c2b21d7dcea3e18ad620ef380ae4fffe04efe1 as its own PR and whoever merges this PR can just squash-merge the whole thing if the history is overly long.

toddrjen commented on scipy/signal/signaltools.py:
```diff
@@ -64,35 +65,39 @@ def _bvalfromboundary(boundary):
-                             "as large as the other in every dimension")
+    for i, (d1, d2) in enumerate(zip(shape1, shape2)):
+        if axes is not None and i not in axes:
+            continue
```
What if `axes` is not defined?

peterbell10 replied:
You could check for that outside the loop:
```python
if axes is None:
	axes = range(len(shape1))
```

toddrjen replied:
I think I have found an even simpler approach.  Please take a look.

toddrjen commented on scipy/signal/signaltools.py:
```diff
@@ -377,58 +546,279 @@ def fftconvolve(in1, in2, mode="full", axes=None):
+        return fallback
+
+    overlap = s2-1
+    opt_size = -overlap*lambertw(-1/(2*math.e*overlap), k=-1).real
```
The link is [here](https://mail.python.org/pipermail/scipy-dev/2019-June/023589.html), but there is an error in the math so I ended up recalculating it myself.  I then validated the results against the "long" form of the equation.

toddrjen replied:
Is this okay now?

peterbell10 replied:
It would be nice to have a reference in the code.

toddrjen commented on scipy/signal/signaltools.py:
```diff
@@ -377,58 +546,279 @@ def fftconvolve(in1, in2, mode="full", axes=None):
+
+    # Reshape the overlap-add parts to input block sizes.
+    split_axes = [iax+i for i, iax in enumerate(axes)]
+    fft_axes = [iax+1 for iax in split_axes]
```
I use `split_axes` in multiple places and I think it would be confusing to recalculate it in some of them, such as line [814](https://github.com/toddrjen/scipy/blob/5fa410c1b81e0e153da240aa9ec383431bb94840/scipy/signal/signaltools.py#L814).

toddrjen replied:
Is this resolved to your satisfaction or do you still think this is an issue?

toddrjen commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -748,62 +763,872 @@ def test_many_sizes(self, n):
+        assert_array_almost_equal(out, expected)
+
+    @pytest.mark.parametrize('axes', [0, [0], -2, [-2]])
+    @pytest.mark.parametrize('shape_a_0', [100, 99, 6, 4])
```
`99` was selected for that reason, and I test a huge number of values in `test_real_manylens`.  Do you have other values you think I should test?  I don't want to add too many since it causes the testing time to grow considerably.

peterbell10 replied:
`99 = 11 * 3 * 3` is now a fast length for the complex FFT. 47 should be good now though.

toddrjen commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -748,62 +763,872 @@ def test_many_sizes(self, n):
+
+        assert_array_almost_equal(out, expected)
+
+    @pytest.mark.parametrize('axes', [1, [1], -1, [-1]])
```
I will move this into a separate set of tests, but it does need to be checked somewhere.  It is surprisingly complicated and fragile.

peterbell10 replied:
`_init_nd_shape_and_axes` is already tested separately. Having a small test case in `TestAllFreqConvolves` would be good though.

toddrjen replied:
Is this okay now?  Rewriting the `fftconvolve` tests are outside the scope of this PR currently but if it is really important I could do it.

toddrjen commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -748,62 +763,872 @@ def test_many_sizes(self, n):
+    @pytest.mark.parametrize('shape_b_0', [100, 99, 6, 4])
+    @pytest.mark.parametrize('shape_a_extra', [1, 3])
+    @pytest.mark.parametrize('shape_b_extra', [1, 3])
+    def test_complex_axes1(self, axes, shape_a_0, shape_b_0,
```
They use a different axes order.  Using parameterize more would make the tests themselves more complicated, making it less clear what they are doing and making it easier to miss important conditions (I have seen conditions missed that way).  I could also combine multiple tests in one method, but that would make the results harder to interpret.  So I was trying to balance length and clarity.  But if the consensus is that this isn't the best approach I could change it.

toddrjen commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -748,62 +763,872 @@ def test_many_sizes(self, n):
+        if shape_a_0 == shape_b_0:
+            return
+        if abs(shape_a_0-shape_b_0) <= 2:
+            return
```
Sure, I will do this.

toddrjen commented:
@peterbell10 I will work on the benchmarks as soon as I can.

toddrjen commented:
@peterbell10 Is there somewhere in the scipy code base I can put the benchmarks?  That might make it easier to look at, but I don't know where that place would be.

peterbell10 commented:
>@peterbell10 Is there somewhere in the scipy code base I can put the benchmarks?

The only thing I'm aware of is the `adv` benchmarks in the `benchmark` folder.  It might be easier to put your code in a gist and link it here though.

toddrjen commented:
Sorry for the delay, I have been trying to figure out the best way to represent the results in an understandable manner.  Here is the ratio of the FFT processing time to the Overlap add processing time (so `FFT/OA`).  Values above 1 mean that that FFT approach took longer, or equivalently that the OA method was faster.  Values with dots are values where the FFT fallback was used, which is necessarily slower than using the FFT approach directly.  

![conv_comparison_perf_ratio](https://user-images.githubusercontent.com/2272878/66408801-9f2b0c00-e9bd-11e9-9556-2640bb46c576.png)

toddrjen commented:
This is the same as above, but using the difference in time rather than the ratio (so `FFT-OA`).  Again, larger values means the OA approach was faster.

![conv_comparison_perf_diff](https://user-images.githubusercontent.com/2272878/66409064-0779ed80-e9be-11e9-807c-29fdb77d2430.png)

![conv_comparison_perf_diff_single](https://user-images.githubusercontent.com/2272878/66409247-5162d380-e9be-11e9-9b32-3ac149ef0d53.png)

toddrjen commented:
Here is the performance ratio again, but the x-axis isn't the size of the first input, but instead is the ratio of the size of the first input to the size of the second input (`in1/in2`).  This shows that the ratio of the two input sizes are the primary determinant of performance.

![conv_comparison_perf_ratio_sz_ratio](https://user-images.githubusercontent.com/2272878/66409634-d0580c00-e9be-11e9-80f0-a33320804b19.png)

![conv_comparison_perf_ratio_sz_ratio_single](https://user-images.githubusercontent.com/2272878/66409648-d2ba6600-e9be-11e9-870f-0e05767dddb2.png)

toddrjen commented:
Here the x-axis is the ratio and the y-axis is the difference.

![conv_comparison_perf_diff_sz_ratio](https://user-images.githubusercontent.com/2272878/66409759-f9789c80-e9be-11e9-88d7-a4c318baffb4.png)

![conv_comparison_perf_diff_sz_ratio_single](https://user-images.githubusercontent.com/2272878/66409767-fbdaf680-e9be-11e9-8c39-dc83e439ffac.png)

peterbell10 commented:
The "FFT-OA" plots might be better on a [symlogy scale](https://matplotlib.org/3.1.1/gallery/scales/symlog_demo.html).

Also, are these benchmarks for 1D arrays?

toddrjen commented:
@peterbell10 

Yes, these are all 1D.  Do you want 2D too?

I am trying out the symlog scale now.  I will post updated figures if I can get it to work.

peterbell10 commented:
> Yes, these are all 1D. Do you want 2D too?

I would be interested to see how the speedup compares in higher dimensions. Maybe not essential though.

toddrjen commented:
I am running the 2D version but it will take time.  Note that doing it in 2D will drastically reduce the maximum size along each axis.  In the meantime, here are the symlog plots:

Input 1 size on x:

![conv_comparison_perf_diff](https://user-images.githubusercontent.com/2272878/66434832-63aa3500-e9f1-11e9-8f9d-c0db3cd177f6.png)

![conv_comparison_perf_diff_single](https://user-images.githubusercontent.com/2272878/66434848-6e64ca00-e9f1-11e9-8c87-d9129e9cbce9.png)

Input size ratio on x:

![conv_comparison_perf_diff_sz_ratio](https://user-images.githubusercontent.com/2272878/66434864-791f5f00-e9f1-11e9-9921-cf661039662f.png)

![conv_comparison_perf_diff_sz_ratio_single](https://user-images.githubusercontent.com/2272878/66434880-84728a80-e9f1-11e9-9ef5-57c00d21507f.png)

toddrjen commented:
Here are the corresponding 2D results, using square arrays.  Note that we aren't getting out to as many samples since my computer is running out of memory, so we aren't seeing the full benefit of the approach.  Sizes are the shape along 1 dimension, not the overall array size.

Performance Ratio vs Input 1 Size

![2D_conv_comparison_perf_ratio](https://user-images.githubusercontent.com/2272878/66656604-54530380-ec0c-11e9-9cab-a16e7a8d06bf.png)

![2D_conv_comparison_perf_ratio_single](https://user-images.githubusercontent.com/2272878/66656623-5d43d500-ec0c-11e9-8ee9-8d15b4fe5d9b.png)


Performance Ratio vs Input Size Ratio

![2D_conv_comparison_perf_ratio_sz_ratio](https://user-images.githubusercontent.com/2272878/66656648-67fe6a00-ec0c-11e9-8000-021ed09d8a75.png)

![2D_conv_comparison_perf_ratio_sz_ratio_single](https://user-images.githubusercontent.com/2272878/66656732-92502780-ec0c-11e9-9ff0-71432858f23c.png)


Performance Difference vs Input 1 Size

![2D_conv_comparison_perf_diff](https://user-images.githubusercontent.com/2272878/66656750-97ad7200-ec0c-11e9-867d-0a545d516f8c.png)

![2D_conv_comparison_perf_diff_single](https://user-images.githubusercontent.com/2272878/66656795-ac8a0580-ec0c-11e9-9af5-18e773516c14.png)



Performance Ratio vs Input Size Difference

![2D_conv_comparison_perf_diff_sz_diff](https://user-images.githubusercontent.com/2272878/66656819-b875c780-ec0c-11e9-80cd-c92e54a39830.png)

![2D_conv_comparison_perf_diff_sz_diff_single](https://user-images.githubusercontent.com/2272878/66656830-bc094e80-ec0c-11e9-9e00-6101c1b156c7.png)

larsoner commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -5,6 +5,7 @@
 
 from decimal import Decimal
 from itertools import product
+from unittest.mock import patch
```
You no longer need this import, right?

toddrjen replied:
Fixed

larsoner commented:
Thanks @toddrjen !
